File: 84.largest-rectangle-in-histogram.py
#
# @lc app=leetcode id=84 lang=python3
#
# [84] Largest Rectangle in Histogram
#

# @lc code=start
from collections import deque


class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        mxrea, s = 0, deque()
        for r, h in enumerate(heights):
            l = r 
            while s and h < s[-1][1]: # previous rec cant extend to current
                l, h0 = s.pop()
                mxrea = max(mxrea, h0 * (r - l))
            s.append((l, h)) # l is lestmost >= h

        return max(mxrea, max(h * (len(heights) - l) for l, h in s))


# @lc code=end

File: 84.largest-rectangle-in-histogram.rs
/*
 * @lc app=leetcode id=84 lang=rust
 *
 * [84] Largest Rectangle in Histogram
 */

// @lc code=start
use std::collections::VecDeque;

impl Solution {
    pub fn largest_rectangle_area(heights: Vec<i32>) -> i32 {
        let (mut mxar, mut s) = (0, VecDeque::new());
        for (r, &h) in heights.iter().enumerate() {
            let mut l = r;

            while s.back().map_or(false, |&(_, h0)| h0 > h) {
                let (l0, h0) = s.pop_back().unwrap();
                mxar = mxar.max(h0 * (r - l0) as i32);
                l = l0;
            }
            s.push_back((l, h));
        }

        let n = heights.len();
        mxar.max(
            s.iter()
                .fold(0, |mx, &(l0, h0)| mx.max(h0 * (n - l0) as i32)),
        )
    }
}
// @lc code=end
